### PR TITLE
REST and WebSocket support for Binance Futures

### PR DESCRIPTION
**Pull request by trexnix**
- Add a REST APIs client for Binance Futures with some basic actions: `ping`, `get_server_time`, `get_exchange_info`, `create_listen_key`, `keep_alive_listen_key`, `get_depth` and `get_account`.

- Add another WebSocket client for Binance Futures

References: https://binanceapitest.github.io/Binance-Futures-API-doc

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -26,8 +26,9 @@ defmodule Binance.MixProject do
   defp deps do
     [
       {:httpoison, "~> 1.0"},
-      {:poison, "~> 4.0.0"},
```
Comment (sirius248) [inline mix.exs:29]: Wondering why we need to downgrade here

Comment (trexnix) [reply mix.exs:29]: This is not final code. There is a dependency conflict, I temporarily downgrade it to make it work so I could focus on implementing the main thing. Will fix it soon.

File: lib/binance/futures.ex [hunk 2]
```diff
@@ -0,0 +1,666 @@
+defmodule Binance.Futures do
+  alias Binance.Futures.Rest.HTTPClient
+
+  # Server
+
+  @doc """
+  Pings binance API. Returns `{:ok, %{}}` if successful, `{:error, reason}` otherwise
+  """
+  def ping() do
+    HTTPClient.get_binance("/fapi/v1/ping")
```
Comment (sirius248) [inline lib/binance/futures.ex:10]: Let move this `/fapi/v1/` into base

File: lib/binance/futures.ex [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+  In the case of a error on binance, for example with invalid parameters, `{:error, {:binance_error, %{code: code, msg: msg}}}` will be returned.
+
+  Please read https://github.com/binance-exchange/binance-official-api-docs/blob/master/rest-api.md#account-information-user_data to understand API
+  """
+
+  def get_account() do
+    api_key = Application.get_env(:binance, :api_key)
+    secret_key = Application.get_env(:binance, :secret_key)
+
+    case HTTPClient.get_binance("/fapi/v3/account", %{}, secret_key, api_key) do
```
Comment (sirius248) [inline lib/binance/futures.ex]: Hum, I see the mix of `v1` and `v3` in this code

File: lib/binance/futures/rest/http_client.ex [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      |> Enum.join("&")
+
+    # generate signature
+    signature =
+      :crypto.hmac(
+        :sha256,
+        Application.get_env(:binance, :secret_key),
+        argument_string
+      )
+      |> Base.encode16()
```
Comment (sirius248) [inline lib/binance/futures/rest/http_client.ex]: Can we refactor this `signing` into it own function.

File: lib/binance/futures/rest/http_client.ex [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        end
+    end
+  end
+
+  def put_binance(url, params) do
+    argument_string =
+      params
+      |> Map.to_list()
+      |> Enum.map(fn x -> Tuple.to_list(x) |> Enum.join("=") end)
+      |> Enum.join("&")
```
Comment (sirius248) [inline lib/binance/futures/rest/http_client.ex]: This parsing part duplicate in all of the functions. Can we extract that out.

File: lib/binance/futures/rest/http_client.ex [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+  def get_binance(url, headers \\ []) do
+    HTTPoison.get("#{@endpoint}#{url}", headers)
+    |> parse_response
+  end
+
+  def delete_binance(url, headers \\ []) do
+    HTTPoison.delete("#{@endpoint}#{url}", headers)
+    |> parse_response
+  end
```
Comment (sirius248) [inline lib/binance/futures/rest/http_client.ex:14]: If use pipe, can we use pipe all the way. It's the convention

```elixir
"#{@endpoint}#{url}"
|> HTTPoison.delete(headers)
|> parse_response
```

Please apply this style to all other places also. We can apply credo for this later on

File: lib/binance/futures/websocket/ws_client.ex [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+  import Logger, only: [info: 1, warn: 1]
+  import Process, only: [send_after: 3]
+
+  # Client API
+  defmacro __using__(_opts) do
+    quote do
+      use WebSockex
+      alias ExOkex.Config
+      @base Application.get_env(:binance, :ws_endpoint, "wss://fstream.binance.com")
+      @ping_interval Application.get_env(:ex_okex, :ping_interval, 5_000)
```
Comment (sirius248) [inline lib/binance/futures/websocket/ws_client.ex]: `ex_okex` -> `:binance`

**Comment by sirius248** [comment]
@trexnix can you add tests for the Future also

File: lib/binance/futures/rest/http_client.ex (hunk 4 above)
Comment (trexnix) [reply lib/binance/futures/rest/http_client.ex]: 👌

File: lib/binance/futures/rest/http_client.ex (hunk 5 above)
Comment (trexnix) [reply lib/binance/futures/rest/http_client.ex]: 👌

File: lib/binance/futures/rest/http_client.ex (hunk 6 above)
Comment (trexnix) [reply lib/binance/futures/rest/http_client.ex:14]: 👍

File: lib/binance/futures/websocket/ws_client.ex (hunk 7 above)
Comment (trexnix) [reply lib/binance/futures/websocket/ws_client.ex]: 👍

**Comment by trexnix** [comment]
> @trexnix can you add tests for the Future also

I added some tests @kimquy

**Comment by sirius248** [comment]
LGTM

**Pull request merged**